File: python/EnigmaProxyClient.py
import requests
from typing import Optional


class EnigmaProxyClient:

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://enigmaproxy.net"
    ):

        self.base_url = base_url.rstrip("/")
        self.session = requests.Session()

        self.session.headers.update({
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json"
        })

    def _request(
        self,
        method: str,
        endpoint: str,
        **kwargs
    ):

        response = self.session.request(
            method=method,
            url=f"{self.base_url}{endpoint}",
            timeout=30,
            **kwargs
        )

        response.raise_for_status()

        return response.json()

    # ----------------------------
    # Packages
    # ----------------------------

    def get_packages(self):
        """
        Get all customer packages.
        """

        return self._request(
            "GET",
            "/api/customer/packages"
        )

    def get_package(
        self,
        package_id: str
    ):
        """
        Get information about a package.
        """

        return self._request(
            "GET",
            f"/api/customer/packages/{package_id}"
        )

    # ----------------------------
    # Proxy Generator
    # ----------------------------

    def generate_proxy(

        self,

        package_id: str,

        protocol: str = "http",

        format: str = "host:port:username:password",

        country: Optional[str] = None,

        state: Optional[str] = None,

        city: Optional[str] = None,

        session: bool = False,

        session_time: Optional[int] = None,

        qty: int = 1,

        lifetime: Optional[int] = None,

        fast_mode: bool = False,

        http3: bool = False

    ):

        payload = {

            "packageId": package_id,

            "protocol": protocol,

            "format": format,

            "qty": qty,

            "session": session,

            "fastMode": fast_mode,

            "http3": http3

        }

        if country:
            payload["country"] = country

        if state:
            payload["state"] = state

        if city:
            payload["city"] = city

        if session_time is not None:
            payload["sessionTime"] = session_time

        if lifetime is not None:
            payload["lifetime"] = lifetime

        return self._request(
            "POST",
            "/api/customer/proxy",
            json=payload
        )

File: go/EnigmaProxyClient.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type EnigmaProxyClient struct {
	APIKey     string
	BaseURL    string
	HTTPClient *http.Client
}

// New creates a new client. An optional baseURL can be passed to override
// the default "https://enigmaproxy.net".
func New(apiKey string, baseURL ...string) *EnigmaProxyClient {

	url := "https://enigmaproxy.net"

	if len(baseURL) > 0 && baseURL[0] != "" {

		url = baseURL[0]

	}

	return &EnigmaProxyClient{

		APIKey: apiKey,

		BaseURL: strings.TrimRight(url, "/"),

		HTTPClient: &http.Client{

			Timeout: 30 * time.Second,
		},
	}
}

func (c *EnigmaProxyClient) request(

	method string,

	endpoint string,

	body interface{},

) (any, error) {

	var reader io.Reader

	if body != nil {

		data, err := json.Marshal(body)

		if err != nil {

			return nil, err

		}

		reader = bytes.NewBuffer(data)

	}

	req, err := http.NewRequest(

		method,

		c.BaseURL+endpoint,

		reader,
	)

	if err != nil {

		return nil, err

	}

	req.Header.Set(

		"Authorization",

		"Bearer "+c.APIKey,
	)

	req.Header.Set(

		"Content-Type",

		"application/json",
	)

	resp, err := c.HTTPClient.Do(req)

	if err != nil {

		return nil, err

	}

	defer resp.Body.Close()

	data, _ := io.ReadAll(resp.Body)

	if resp.StatusCode >= 400 {

		return nil,

			fmt.Errorf(

				"request failed (%d): %s",

				resp.StatusCode,

				string(data),
			)

	}

	var result any

	err = json.Unmarshal(

		data,

		&result,
	)

	return result, err

}

// ----------------------------
// Packages
// ----------------------------

func (c *EnigmaProxyClient) GetPackages() (any, error) {

	return c.request(

		"GET",

		"/api/customer/packages",

		nil,
	)

}

func (c *EnigmaProxyClient) GetPackage(

	packageID string,

) (any, error) {

	return c.request(

		"GET",

		"/api/customer/packages/"+packageID,

		nil,
	)

}

// ----------------------------
// Proxy Generator
// ----------------------------

type GenerateProxyOptions struct {
	PackageID string

	Protocol string

	Format string

	Country string

	State string

	City string

	Session bool

	SessionTime *int

	Qty int

	Lifetime *int

	FastMode bool

	HTTP3 bool
}

func (c *EnigmaProxyClient) GenerateProxy(

	opts GenerateProxyOptions,

) (any, error) {

	protocol := opts.Protocol

	if protocol == "" {

		protocol = "http"

	}

	format := opts.Format

	if format == "" {

		format = "host:port:username:password"

	}

	qty := opts.Qty

	if qty == 0 {

		qty = 1

	}

	payload := map[string]interface{}{

		"packageId": opts.PackageID,

		"protocol": protocol,

		"format": format,

		"qty": qty,

		"session": opts.Session,

		"fastMode": opts.FastMode,

		"http3": opts.HTTP3,
	}

	if opts.Country != "" {

		payload["country"] = opts.Country

	}

	if opts.State != "" {

		payload["state"] = opts.State

	}

	if opts.City != "" {

		payload["city"] = opts.City

	}

	if opts.SessionTime != nil {

		payload["sessionTime"] = *opts.SessionTime

	}

	if opts.Lifetime != nil {

		payload["lifetime"] = *opts.Lifetime

	}

	return c.request(

		"POST",

		"/api/customer/proxy",

		payload,
	)

}
